codec/decode.go

```diff
@@ -2413,8 +2413,24 @@ func (d *Decoder) rawBytes() []byte {
 	return bs2
 }
 
-func (d *Decoder) wrapErrstr(v interface{}, err *error) {
-	*err = fmt.Errorf("%s decode error [pos %d]: %v", d.hh.Name(), d.r.numread(), v)
+// DecodeError wraps any error encountered during decoding.
+type DecodeError struct {
+	Name string
+	Pos  int
+	Err  error
+}
+
+func (d DecodeError) Error() string {
+	return fmt.Sprintf("%s decode error [pos %d]: %v", d.Name, d.Pos, d.Err)
+}
+
+// Cause implements github.com/pkg/errors.causer.
+func (d DecodeError) Cause() error {
+	return d.Err
+}
+
```

Comment:
Is there a good reason for exporting the Cause() method? Why?

Reply:
It has to be exported to implement the causer interface